js/components/MilkdropWindow/index.js

```diff
@@ -0,0 +1,111 @@
+import React from "react";
+import screenfull from "screenfull";
+import butterchurn from "butterchurn";
+import reactionDiffusion2 from "butterchurn-presets/presets/converted/Geiss - Reaction Diffusion 2";
+
+class MilkdropWindow extends React.Component {
+  constructor() {
+    super();
+  }
+  componentDidMount() {
+    const scale = window.devicePixelRatio;
+    const analyserNode = this.props.analyser;
+    this.visualizer = butterchurn.createVisualizer(
+      analyserNode.context,
+      this._canvasNode,
+      {
+        width: this._canvasNode.width * scale,
+        height: this._canvasNode.height * scale
+      }
+    );
+    this.visualizer.connectAudio(analyserNode);
+    this.visualizer.loadPreset(reactionDiffusion2);
+    this._renderViz();
+
+    this._canvasNode.addEventListener("dblclick", () => {
+      this._handleRequestFullsceen();
+    });
+
+    require.ensure(["butterchurn-presets"], require => {
+      const butterchurnPresets = require("butterchurn-presets");
+      const presets = butterchurnPresets.getPresets();
+      const presetKeys = Object.keys(presets);
+      this.cycleInterval = setInterval(() => {
+        const presetIdx = Math.floor(presetKeys.length * Math.random());
+        const preset = presets[presetKeys[presetIdx]];
+        this.visualizer.loadPreset(preset);
+      }, 15000);
+    });
+  }
+  componentWillUnmount() {
+    this._pauseViz();
+    this._stopCycling();
+  }
+  componentDidUpdate() {
+    if (
+      !screenfull.isFullscreen &&
+      (this.props.width !== this.width || this.props.height !== this.height)
```

Comment:
Optional nit:

`componentDidUpdate` gets passed `prevProps` as the first argument. Rather than storing the height/width as a direct property, we can just check:

```
(this.props.width !== prevProps.width || this.props.height !== prevProps.height)
```

Reply:
ahhhh, totally forgot about that..been a bit since I did react ha, thanks!

Reply:
No worries. I'm not at all concerned about the framework specific stuff, I just figured I'd point it out. I'm happy to clean these nits up later if you'd rather not be bothered.